core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -733,6 +750,7 @@ sealed trait ZIO[-R, +E, +A] extends Serializable with ZIOPlatformSpecific[R, E,
   /**
    * Supplies `zio` if the predicate fails.
    */
+  // TODO: Should we have a `filterOrElseZIO` and `filterOrElse`?
   final def filterOrElse[R1 <: R, E1 >: E, A1 >: A](p: A => Boolean)(zio: => ZIO[R1, E1, A1]): ZIO[R1, E1, A1] =
```

Comment:
Methods like this already have so many variants I would be inclined to not add more. It is already quite easy to do this but evaluating an effectual condition and then doing something else based on it.